pipeline/authn/authenticator_cookie_session.go

```diff
@@ -171,8 +173,20 @@ func forwardRequestToSessionStore(r *http.Request, checkSessionURL string, prese
 		return nil, helper.ErrForbidden.WithReason(err.Error()).WithTrace(err)
 	}
 
+	defer res.Body.Close()
+
 	if res.StatusCode == 200 {
-		body, err := ioutil.ReadAll(res.Body)
+		var reader io.Reader
+		switch res.Header.Get("Content-Encoding") {
+		case "gzip":
```

Comment:
`http.DefaultTransport` automatically handles gzip decoding if the correct headers are set: https://stackoverflow.com/questions/13130341/reading-gzipped-http-response-in-go

So this should not be required!

Reply:
the `http.DefaultTransport` does not use the `DisableCompression` property, and it's disabled by default. Also, the standard HTTP client does not support automatic decompression of `deflate` encoding. 

I can give you a way how to reproduce the issue because I reproduced it by having this setup.

1. Ory proxy proxies traffic from Ory Cloud
2. Oathkeeper uses ory proxy endpoint and cookie_session authenticator
3. Oathkeeper as one upstream to proxy traffic to
I found that with mutator header `X-User` that comes from the cookie_authenticator is always empty and after a couple of hours of debugging I found that's a gzip issue.

I think that I'll write a couple of tests to prove the bug, and maybe I'll find a better solution, but also a couple of users had the same issue in our Slack community

Reply:
However, the default client[ handles this by itself ](https://go.dev/src/net/http/transport.go#L2196). I think I'll start with the tests first and try to reproduce it on a clean setup

Reply:
Ok, that sounds great! We had issues with GZIP in the past in Ory Oathkeeper. I think it depends a bit on where in the pipeline the issue is. I think the proxy had issues with GZIP as it did not decode them automatically - here however I think we do?

Another problem that can happen is that we forward requests with all the headers that might include e.g. `Content-Encoding: gzip` but we actually do not use gzip in the body. So basically an incoming request to oathkeeper is:

```
GET /foo
Content-Encoding: gzip

...gzip-body...
```

then Ory Oathkeeper reads that data and decompresses it. Next, it forwards the request to e.g. the session endpoint with all the HTTP headers:

```
GET /sessions/whoami
Content-Encoding: gzip

...here the body is actually not gzip any more because it was decoded in oathkeeper already...
```

which then causes issues at the receiver.

I think that, in the end, we will need to refactor this so that we only send "wanted" headers to the upstream services instead of sending a catch-all. The same underlying issues is the reason for this: https://github.com/ory/oathkeeper/pull/954

Reply:
I think the problem is the opposite way around... The behaviour of the Go HTTP client depends on whether the `http.Request` explicitly sets the `Accept-Encoding: gzip` header or not. If no header is set, the default http client automatically adds it before firing the request, and transparently handles decompression of the response for you (stripping out any `Content-Encoding` response headers).

If the request header is set though, it assumes you want to handle the decompression manually, and so returns the response compressed.

This can happen and cause problems if the service you're proxying serves compressed content in response to a request with `Accept-Encoding: gzip` (e.g. an nginx ingress with transparent compression enabled) - Oathkeeper would receive the response which would be decompressed, but because the inbound request had the `Accept-Encoding` header set, the go http client would not transparently decompress it.

The offending lines are below, the blind copy of all inbound request headers onto the outbound request on L161. 
```go
	for k, v := range r.Header {
		req.Header[k] = v
	}
```

We should add a check for the `Accept-Encoding` header here, and not copy it, to avoid this

Reply:
Right, that can also very well be the root cause! I think the effect and best resolution is to only pass essential and configured headers to the upstream and leave the user the ability to pass additional headers through the config too.